README.md
<p align="center">
<a href="https://github.com/spapanik/pyrencode/actions/workflows/build.yml"><img alt="Build" src="https://github.com/spapanik/pyrencode/actions/workflows/build.yml/badge.svg"></a>
<a href="https://lgtm.com/projects/g/spapanik/pyrencode/alerts/"><img alt="Total alerts" src="https://img.shields.io/lgtm/alerts/g/spapanik/pyrencode.svg"/></a>
<a href="https://github.com/spapanik/pyrencode/blob/master/LICENSE.txt"><img alt="License" src="https://img.shields.io/github/license/spapanik/pyrencode"></a>
<a href="https://pypi.org/project/pyrencode"><img alt="PyPI" src="https://img.shields.io/pypi/v/pyrencode"></a>
<a href="https://pepy.tech/project/pyrencode"><img alt="Downloads" src="https://pepy.tech/badge/pyrencode"></a>
<a href="https://github.com/psf/black"><img alt="Code style" src="https://img.shields.io/badge/code%20style-black-000000.svg"></a>
</p>

# pyrencode

_pyrencode_ is a python rewrite of the original rencode by aresch (https://github.com/aresch/rencode), avoiding  the cython dependency, so that it can be used with PyPy.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][clog], and this project adheres to [Semantic Versioning][semver].

## [Unreleased]

### Added
- Added python310 support
- Added a changelog

## [0.1.4] - 2018-10-06

## [0.1.3] - 2018-10-06

## [0.0.5] - 2017-08-05

## [0.0.4] - 2016-09-26

## [0.0.3] - 2016-09-25

## [0.0.2] - 2016-07-25

## [0.0.1] - 2016-06-08


[unreleased]: https://github.com/spapanik/pyrencode/compare/v0.1.4...master
[0.1.4]: https://github.com/spapanik/pyrencode/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/spapanik/pyrencode/compare/v0.0.5...v0.1.3
[0.0.5]: https://github.com/spapanik/pyrencode/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/spapanik/pyrencode/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/spapanik/pyrencode/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/spapanik/pyrencode/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/spapanik/pyrencode/releases/tag/v0.0.1

[clog]: https://keepachangelog.com/en/1.0.0/
[semver]: https://semver.org/spec/v2.0.0.html


tests/test_roundtrip.py
from pyrencode.encode import dumps
from pyrencode.decode import loads

import struct

f1 = struct.unpack('!f', struct.pack('!f', 25.5))[0]
f2 = struct.unpack('!f', struct.pack('!f', 29.3))[0]
f3 = struct.unpack('!f', struct.pack('!f', -0.6))[0]


def test_mixed_types_tuple():
    obj = (
        (
            {
                b'a': 15,
                b'bb': f1,
                b'ccc': f2,
                b'': (f3, (), False, True, b'')
            },
            (b'a', 10 ** 20), tuple(range(-100000, 100000)),
            b'b' * 31, b'b' * 62, b'b' * 64, 2 ** 30,
            2 ** 33, 2 ** 62, 2 ** 64, 2 ** 30, 2 ** 33, 2 ** 62, 2 ** 64,
            False, False, True,
            -1, 2, 0,
            None,
        ),
    )
    assert loads(dumps(obj)) == obj


def test_tuple_of_updated_dict():
    d = dict(zip(range(-100000, 100000), range(-100000, 100000)))
    d.update(
        {
            b'a': 20, 20: 40, 40: 41,
            f1: f2, f2: f3, f3: False,
            False: True,
            True: False,
        }
    )
    obj = (d, {}, {5: 6}, {7: 7, True: 8}, {9: 10, 22: 39, 49: 50, 44: b''})
    assert loads(dumps(obj)) == obj


def test_tuple_of_bytes():
    obj = (
        b'', b'a' * 10, b'a' * 100, b'a' * 1000, b'a' * 10000,
        b'a' * 100000, b'a' * 1000000, b'a' * 10000000
    ) + tuple([b'a' * n for n in range(1000)]) + (b'b',)
    assert loads(dumps(obj)) == obj


def test_tuple_of_dicts():
    obj = tuple(
        [dict(zip(range(n), range(n))) for n in range(100)]
    ) + tuple(
        [dict(zip(range(n), range(-n, 0))) for n in range(100)]
    ) + (b'b',)
    assert loads(dumps(obj)) == obj


def test_tuple_of_tuples():
    obj = tuple([tuple(range(n)) for n in range(100)]) + (b'b',)
    assert loads(dumps(obj)) == obj


def test_tuple_of_bool():
    obj = (None, True, None, False)
    assert loads(dumps(obj)) == obj


def test_none():
    assert loads(dumps(None)) is None


def test_none_dict():
    assert loads(dumps({None: None})) == {None: None}


def test_float():
    assert 1e-10 < abs(loads(dumps(1.1)) - 1.1) < 1e-6


def test_float_32():
    assert 1e-10 < abs(loads(dumps(1.1, 32)) - 1.1) < 1e-6


def test_float_64():
    assert abs(loads(dumps(1.1, 64)) - 1.1) < 1e-12


def test_string():
    assert loads(dumps("Hello World!!"), decode_utf8=True) == "Hello World!!"


tests/pyrencode/test_encode.py
import pytest

from pyrencode import encode


def test_encode_int_under_max_size():
    encode.encode_int(2 ** 128, [])


def test_encode_int_over_max_size():
    pytest.raises(OverflowError, encode.encode_int, 2 ** 256, [])


def test_dumps_wrong_float_bits():
    pytest.raises(ValueError, encode.dumps, None, float_bits=128)


src/pyrencode/decode.py
import struct

from pyrencode.settings import *

_decode_utf8 = DECODE_UTF8


def decode_int(bytes_obj, cursor):
    cursor += 1
    new_cursor = bytes_obj.index(CHR_TERM, cursor)
    if new_cursor - cursor >= MAX_INT_LENGTH:
        raise ValueError("overflow")
    n = int(bytes_obj[cursor:new_cursor])
    if bytes_obj[cursor : cursor + 1] == "-":
        if bytes_obj[cursor + 1 : cursor + 2] == "0":
            raise ValueError
    elif bytes_obj[cursor : cursor + 1] == "0" and new_cursor != cursor + 1:
        raise ValueError

    return n, new_cursor + 1


def decode_int1(bytes_obj, cursor):
    cursor += 1
    return struct.unpack("!b", bytes_obj[cursor : cursor + 1])[0], cursor + 1


def decode_int2(bytes_obj, cursor):
    cursor += 1
    return struct.unpack("!h", bytes_obj[cursor : cursor + 2])[0], cursor + 2


def decode_int3(bytes_obj, cursor):
    cursor += 1
    return struct.unpack("!l", bytes_obj[cursor : cursor + 4])[0], cursor + 4


def decode_int4(bytes_obj, cursor):
    cursor += 1
    return struct.unpack("!q", bytes_obj[cursor : cursor + 8])[0], cursor + 8


def decode_float32(bytes_obj, cursor):
    cursor += 1
    return struct.unpack("!f", bytes_obj[cursor : cursor + 4])[0], cursor + 4


def decode_float64(bytes_obj, cursor):
    cursor += 1
    return struct.unpack("!d", bytes_obj[cursor : cursor + 8])[0], cursor + 8


def decode_string(bytes_obj, cursor):
    colon = bytes_obj.index(b":", cursor)
    n = int(bytes_obj[cursor:colon])
    if bytes_obj[cursor] == "0" and colon != cursor + 1:
        raise ValueError
    colon += 1
    string = bytes_obj[colon : colon + n]
    if _decode_utf8:
        string = string.decode("utf8")
    return string, colon + n


def decode_list(bytes_obj, cursor):
    r, cursor = [], cursor + 1
    while bytes_obj[cursor : cursor + 1] != CHR_TERM:
        v, cursor = decode_func[bytes_obj[cursor : cursor + 1]](bytes_obj, cursor)
        r.append(v)
    return tuple(r), cursor + 1


def decode_dict(x, cursor):
    r, cursor = {}, cursor + 1
    while x[cursor : cursor + 1] != CHR_TERM:
        k, cursor = decode_func[x[cursor : cursor + 1]](x, cursor)
        r[k], cursor = decode_func[x[cursor : cursor + 1]](x, cursor)
    return r, cursor + 1


def decode_true(_, cursor):
    return True, cursor + 1


def decode_false(_, cursor):
    return False, cursor + 1


def decode_none(_, cursor):
    return None, cursor + 1


decode_func = {
    b"0": decode_string,
    b"1": decode_string,
    b"2": decode_string,
    b"3": decode_string,
    b"4": decode_string,
    b"5": decode_string,
    b"6": decode_string,
    b"7": decode_string,
    b"8": decode_string,
    b"9": decode_string,
    CHR_LIST: decode_list,
    CHR_DICT: decode_dict,
    CHR_INT: decode_int,
    CHR_INT1: decode_int1,
    CHR_INT2: decode_int2,
    CHR_INT4: decode_int3,
    CHR_INT8: decode_int4,
    CHR_FLOAT32: decode_float32,
    CHR_FLOAT64: decode_float64,
    CHR_TRUE: decode_true,
    CHR_FALSE: decode_false,
    CHR_NONE: decode_none,
}


def make_fixed_length_string_decoders():
    def make_decoder(slen):
        def func(x, f):
            s = x[f + 1 : f + 1 + slen]
            if _decode_utf8:
                s = s.decode("utf8")
            return s, f + 1 + slen

        return func

    for i in range(STR_FIXED_COUNT):
        decode_func[int2byte(STR_FIXED_START + i)] = make_decoder(i)


make_fixed_length_string_decoders()


def make_fixed_length_list_decoders():
    def make_decoder(slen):
        def func(x, f):
            r, f = [], f + 1
            for _ in range(slen):
                v, f = decode_func[x[f : f + 1]](x, f)
                r.append(v)
            return tuple(r), f

        return func

    for i in range(LIST_FIXED_COUNT):
        decode_func[int2byte(LIST_FIXED_START + i)] = make_decoder(i)


make_fixed_length_list_decoders()


def make_fixed_length_int_decoders():
    def make_decoder(j):
        def func(_, f):
            return j, f + 1

        return func

    for i in range(INT_POS_FIXED_COUNT):
        decode_func[int2byte(INT_POS_FIXED_START + i)] = make_decoder(i)
    for i in range(INT_NEG_FIXED_COUNT):
        decode_func[int2byte(INT_NEG_FIXED_START + i)] = make_decoder(-1 - i)


make_fixed_length_int_decoders()


def make_fixed_length_dict_decoders():
    def make_decoder(slen):
        def func(x, f):
            r, f = {}, f + 1
            for _ in range(slen):
                k, f = decode_func[x[f : f + 1]](x, f)
                r[k], f = decode_func[x[f : f + 1]](x, f)
            return r, f

        return func

    for i in range(DICT_FIXED_COUNT):
        decode_func[int2byte(DICT_FIXED_START + i)] = make_decoder(i)


make_fixed_length_dict_decoders()


def loads(bytes_obj, decode_utf8=False):
    global _decode_utf8
    _decode_utf8 = decode_utf8
    try:
        obj, end_position = decode_func[bytes_obj[0:1]](bytes_obj, 0)
    except (IndexError, KeyError):
        raise ValueError
    if end_position != len(bytes_obj):
        raise ValueError
    return obj


src/pyrencode/encode.py
import struct
from threading import Lock

from pyrencode.settings import *


def encode_int(obj, data_list):
    if 0 <= obj < INT_POS_FIXED_COUNT:
        data_list.append(int2byte(INT_POS_FIXED_START + obj))
    elif -INT_NEG_FIXED_COUNT <= obj < 0:
        data_list.append(int2byte(INT_NEG_FIXED_START - 1 - obj))
    elif -INT1_SIZE <= obj < INT1_SIZE:
        data_list.extend((CHR_INT1, struct.pack("!b", obj)))
    elif -INT2_SIZE <= obj < INT2_SIZE:
        data_list.extend((CHR_INT2, struct.pack("!h", obj)))
    elif -INT4_SIZE <= obj < INT4_SIZE:
        data_list.extend((CHR_INT4, struct.pack("!l", obj)))
    elif -INT8_SIZE <= obj < INT8_SIZE:
        data_list.extend((CHR_INT8, struct.pack("!q", obj)))
    else:
        s = bytes(str(obj), "ascii")
        if len(s) >= MAX_INT_LENGTH:
            raise OverflowError("Integer is too long to be rencoded.")

        data_list.extend((CHR_INT, s, CHR_TERM))


def encode_float32(obj, data_list):
    data_list.extend((CHR_FLOAT32, struct.pack("!f", obj)))


def encode_float64(obj, data_list):
    data_list.extend((CHR_FLOAT64, struct.pack("!d", obj)))


def encode_bool(obj, data_list):
    if obj:
        data_list.append(CHR_TRUE)
    else:
        data_list.append(CHR_FALSE)


def encode_none(_, data_list):
    data_list.append(CHR_NONE)


def encode_bytes(obj, data_list):
    if len(obj) < STR_FIXED_COUNT:
        data_list.extend((int2byte(STR_FIXED_START + len(obj)), obj))
    else:
        string = bytes(str(len(obj)), ASCII)
        data_list.extend((string, b":", obj))


def encode_string(obj, data_list):
    encode_bytes(obj.encode(UTF8), data_list)


def encode_list(obj, data_list):
    if len(obj) < LIST_FIXED_COUNT:
        data_list.append(int2byte(LIST_FIXED_START + len(obj)))
        for item in obj:
            encode_func[type(item)](item, data_list)
    else:
        data_list.append(CHR_LIST)
        for item in obj:
            encode_func[type(item)](item, data_list)
        data_list.append(CHR_TERM)


def encode_dict(obj, data_list):
    if len(obj) < DICT_FIXED_COUNT:
        data_list.append(int2byte(DICT_FIXED_START + len(obj)))
        for key, value in obj.items():
            encode_func[type(key)](key, data_list)
            encode_func[type(value)](value, data_list)
    else:
        data_list.append(CHR_DICT)
        for key, value in obj.items():
            encode_func[type(key)](key, data_list)
            encode_func[type(value)](value, data_list)
        data_list.append(CHR_TERM)


encode_func = {
    int: encode_int,
    bytes: encode_bytes,
    list: encode_list,
    tuple: encode_list,
    dict: encode_dict,
    type(None): encode_none,
    str: encode_string,
    bool: encode_bool,
}

lock = Lock()


def dumps(obj, float_bits=DEFAULT_FLOAT_BITS):
    data_list = None
    with lock:
        if float_bits == 32:
            encode_func[float] = encode_float32
        elif float_bits == 64:
            encode_func[float] = encode_float64
        else:
            raise ValueError(
                "Float bits {float_bits} is not 32 or 64".format(float_bits=float_bits)
            )
        data_list = []
        encode_func[type(obj)](obj, data_list)

    return b"".join(data_list)
